### DISCUSSION TITLE
MCP Gateway: bridge STDIO MCP servers to HTTP+SSE for remote access

### DISCUSSION THREAD

**Discussion by acehoss** [danny-avila/LibreChat#5032] [General]
I've been enjoying the brand new MCP tools feature in LibreChat, and I found myself wanting to use some of the really interesting STDIO MCP servers that are out there. While LibreChat does support STDIO servers directly, running them in the container can be tricky - you need all the right dependencies installed, and you might not have access to the resources that make the server useful (like the local filesystem).

I looked around for a solution to bridge STDIO servers to HTTP+SSE so I could run them outside the container, but couldn't find anything that quite fit. So I built a simple gateway that:
- Bridges STDIO MCP servers to HTTP+SSE transport
- Lets you run multiple server endpoints through a single gateway
- Uses simple YAML configuration

You can find it here: [MCP Gateway](https://github.com/acehoss/mcp-gateway)

### Warning
⚠️ MCP Gateway doesn't have any authentication or authorization yet. Please only run it on trusted networks!

**Comment by danny-avila** [comment]
There's only one example of an MCP Server using SSE Transport: https://github.com/modelcontextprotocol/servers/blob/main/src/everything/sse.ts

You could follow that example, but would require editing existing servers' code and/or developing your own from scratch.

Anthropic dev on this, seems SSE Transport is still a work in progress, and maybe why no one is really iterating them:
> We're definitely interested in extending MCP to cover remote connections as well. Both SDKs already support an SSE transport with that in mind: [https://modelcontextprotocol.io/docs/concepts/transports#ser...](https://modelcontextprotocol.io/docs/concepts/transports#server-sent-events-sse)

> However, it's not quite a complete story yet. Remote connections introduce a lot more questions and complexity—related to deployment, auth, security, etc. We'll be working through these in the coming weeks, and would love any and all input!

source: https://news.ycombinator.com/item?id=42237424

**Reply by acehoss** [reply]
Heh, yeah, I saw that post while looking for documentation on the protocol, and there definitely a lot of "needs attention" aspects on the network-based MCP. Still, it was a good exercise, and it does work with LibreChat.

In retrospect, I could have accomplished everything I wanted with SSH. Keys and such would have to be mapped into the container, but I tried using `ssh -t` as a bridge to an MCP server on another computer with Claude Desktop and it worked great.

**Reply by danny-avila** [reply]
> We'll soon provide developer toolkits for deploying remote production MCP servers

they also mentioned it here: https://www.anthropic.com/news/model-context-protocol

**Comment by danny-avila** [comment]
[MCP Gateway](https://github.com/acehoss/mcp-gateway) looks cool, I would try to share it in the [https://github.com/modelcontextprotocol/servers](https://github.com/modelcontextprotocol/servers) repo

**Comment by aleibovici** [comment]
Could you provide an example of how the mcp-gateway should be configured in Librechat?

**Reply by acehoss** [reply]
There are some examples on this issue: https://github.com/acehoss/mcp-gateway/issues/1

**Reply by aleibovici** [reply]
Thanks
